src/d3d12/d3d12_resource_pool_model.cpp

```diff
@@ -0,0 +1,398 @@
+			sizeof(std::uint32_t),
+			ResourceState::INDEX_BUFFER);
+
+		m_index_buffer_size = index_buffer_size_in_mb * 1024 * 1024;
+
+		auto page_frame_count = SizeAlign(m_vertex_buffer_size / 65536, 64) / 64;
+		auto frame_count = m_vertex_buffer_size / 65536;
+
+		m_vertex_buffer_bitmap.resize(page_frame_count);
+
+		for (int i = 0; i < m_vertex_buffer_bitmap.size() - 1; ++i) {
+			m_vertex_buffer_bitmap[i] = 0xffffffffffffffff;
+		}
+		for (int i = 0; i < 64; ++i) {
+			if ((m_vertex_buffer_bitmap.size() - 1) * 64 + i < frame_count) {
+				m_vertex_buffer_bitmap[m_vertex_buffer_bitmap.size() - 1] |= 1Ui64 << i;
+			}
+			else {
+				break;
+			}
+		}
+		
+		page_frame_count = SizeAlign(m_index_buffer_size / 65536, 64) / 64;
+		frame_count = m_index_buffer_size / 65536;
+
+		m_index_buffer_bitmap.resize(page_frame_count);
+
+		for (int i = 0; i < m_index_buffer_bitmap.size() - 1; ++i) {
+			m_index_buffer_bitmap[i] = 0xffffffffffffffff;
+		}
+		for (int i = 0; i < 64; ++i) {
+			if ((m_index_buffer_bitmap.size() - 1) * 64 + i < frame_count) {
+				m_index_buffer_bitmap[m_index_buffer_bitmap.size() - 1] |= 1Ui64 << i;
+			}
+			else {
+				break;
+			}
+		}
+	}
+
+	D3D12ModelPool::~D3D12ModelPool()
+	{
+		d3d12::Destroy(m_vertex_buffer);
+		d3d12::Destroy(m_index_buffer);
+
+		//Do model handles created from this pool be destroyed by the pool?
+	}
+
+	void D3D12ModelPool::Evict()
+	{
+		d3d12::Evict(m_vertex_buffer);
+		d3d12::Evict(m_index_buffer);
+	}
+
+	void D3D12ModelPool::MakeResident()
+	{
+		d3d12::MakeResident(m_vertex_buffer);
+		d3d12::MakeResident(m_index_buffer);
+	}
+
+	void D3D12ModelPool::StageMesh(Mesh * mesh, d3d12::CommandList * cmd_list)
+	{
+		D3D12Mesh* d3d12_mesh = static_cast<D3D12Mesh*>(mesh);
+		d3d12::StageBufferRegion(m_vertex_buffer, 
+			d3d12_mesh->m_vertex_staging_buffer_size, 
+			d3d12_mesh->m_vertex_staging_buffer_offset, 
+			cmd_list);
+		d3d12::StageBufferRegion(m_index_buffer,
+			d3d12_mesh->m_index_staging_buffer_size,
+			d3d12_mesh->m_index_staging_buffer_offset,
+			cmd_list);
+	}
+
+	d3d12::StagingBuffer * D3D12ModelPool::GetVertexStagingBuffer()
+	{
+		return m_vertex_buffer;
+	}
+
+	d3d12::StagingBuffer * D3D12ModelPool::GetIndexStagingBuffer()
+	{
+		return m_index_buffer;
+	}
+
+	Model* D3D12ModelPool::LoadFBX(std::string_view path, ModelType type)
+	{
+		return new Model();
+	}
+
+	Mesh* D3D12ModelPool::LoadCustom_VerticesAndIndices(void* vertices_data, std::size_t num_vertices, std::size_t vertex_size, void* indices_data, std::size_t num_indices, std::size_t index_size)
+	{
+		D3D12Mesh* mesh = new D3D12Mesh();
+		memset(mesh, 0, sizeof(D3D12Mesh));
+		mesh->m_model_pool = this;
+
+		auto frame_count = m_vertex_buffer_size / 65536;
+		auto needed_frames = SizeAlign(num_vertices*vertex_size, 65536) / 65536;
+
+		std::uint64_t start_frame = 0;
+		bool counting = false;
+		bool found = false;
+		int free_frames = 0;
+
+		for (std::uint64_t i = 0; i <= IndexFromBit(frame_count); ++i) {
```

Comment:
Please explain what is going on here. The code is quite ugly. I'm not sure how you would structure this better since I'm not 100% sure whats happening.